### PR TITLE
Add `qr` decomposition for `Tensor`s

### LINKED ISSUE #2: Implement QR factorization
@HerManNav @jofrevalles can you implement this?

### PR DESCRIPTION
**Pull request by jofrevalles**
### Summary
This pull request addresses issue "Implement QR factorization" (resolve #2) by adding the `qr` function for `Tensor`s, based on the full QR decomposition provided by the `LinearAlgebra` library. The implementation includes the necessary error handling and dimension reshaping to work seamlessly with Tensors. In addition, tests have been added to ensure the accuracy and correctness of the new function.

### Example
```julia
julia> using Tensors
julia> using OMEinsum
julia> using LinearAlgebra
julia> using Test

julia> data = rand(2, 4, 6, 8)
2×4×6×8 Array{Float64, 4}:
...

julia> tensor = Tensor(data, (:i, :j, :k, :l))
2×4×6×8 Tensor{Float64, 4, Array{Float64, 4}}:
...

julia> Q, R = qr(tensor, labels(tensor)[1:2])
...

julia> tensor_recovered = ein"ijk,klm->ijlm"(Q, R)
2×4×6×8 Array{Float64, 4}:
...

julia> @test tensor_recovered ≈ tensor
Test Passed
```

### CODE REVIEW

**Comment by jofrevalles** [comment]
Note that the tests fail now because we use the function introduced in https://github.com/bsc-quantic/Tensors.jl/pull/20 that is still not in `master`. With this function, the tests work as expected.

File: src/Numerics.jl [hunk 1]
```diff
@@ -71,3 +71,38 @@ function LinearAlgebra.svd(t::Tensor, left_inds; kwargs...)
...
+    elseif any(∉(labels(t)), left_inds)
+        # TODO better error exception and checks
+        throw(ErrorException("all left-indices must be in $(labels(t))"))
+    end
+
+    right_inds = setdiff(labels(t), left_inds)
+    if isempty(right_inds)
+        # TODO better error exception and checks
+        throw(ErrorException("no right-indices in QR factorization"))
+    end
```
Comment (mofeing) [inline src/Numerics.jl:89]:
```suggestion
    # TODO better error exception and checks
    isempty(left_inds) && throw(ErrorException("no left-indices in QR factorization"))
    left_inds ⊆ labels(t) || throw(ErrorException("all left-indices must be in $(labels(t))"))

    right_inds = setdiff(labels(t), left_inds)
    isempty(right_inds) && throw(ErrorException("no right-indices in QR factorization"))
```

File: test/Numerics_test.jl [hunk 2]
```diff
@@ -127,4 +127,50 @@
...
+        tensor = Tensor(data, (:i, :j, :k))
+
+        @testset "Error Handling Test" begin
+            # Throw exception if left_inds is not provided
+            @test_throws ErrorException qr(tensor)
+            # Throw exception if left_inds ∉ labels(tensor)
+            @test_throws ErrorException qr(tensor, (:l,))
+        end
+
+        @testset "Labels Test" begin
```
Comment (mofeing) [inline test/Numerics_test.jl:142]:
```suggestion
        @testset "labels" begin
```

File: test/Numerics_test.jl [hunk 3]
```diff
@@ -127,4 +127,50 @@
             end
         end
     end
+
+    @testset "qr" begin
+        data = rand(2, 2, 2)
+        tensor = Tensor(data, (:i, :j, :k))
+
+        @testset "Error Handling Test" begin
```
Comment (mofeing) [inline test/Numerics_test.jl:135]:
```suggestion
        @testset "[exceptions]" begin
```

File: test/Numerics_test.jl [hunk 4]
```diff
@@ -127,4 +127,50 @@
...
+            # Throw exception if left_inds is not provided
+            @test_throws ErrorException qr(tensor)
+            # Throw exception if left_inds ∉ labels(tensor)
+            @test_throws ErrorException qr(tensor, (:l,))
```
Comment (mofeing) [inline test/Numerics_test.jl:139]:
```suggestion
            @test_throws ErrorException qr(tensor, (:l,))
            # throw exception if no right-inds
            @test_throws ErrorException qr(tensor, (:i,:j,:k))
```

File: test/Numerics_test.jl [hunk 5]
```diff
@@ -127,4 +127,50 @@
...
+        end
+
+        @testset "Labels Test" begin
+            Q, R = qr(tensor, labels(tensor)[1:2])
+            @test labels(Q)[1:2] == labels(tensor)[1:2]
+            @test labels(Q)[3] == labels(R)[1]
+            @test labels(R)[2] == labels(tensor)[3]
+        end
+
+        @testset "Size Test" begin
```
Comment (mofeing) [inline test/Numerics_test.jl:149]:
```suggestion
        @testset "size" begin
```

File: test/Numerics_test.jl [hunk 6]
```diff
@@ -127,4 +127,50 @@
...
+
+            # Additional test with different dimensions
+            data2 = rand(2, 4, 6, 8)
+            tensor2 = Tensor(data2, (:i, :j, :k, :l))
+            Q2, R2 = qr(tensor2, labels(tensor2)[1:2])
+            @test size(Q2) == (2, 4, 8)
+            @test size(R2) == (8, 6, 8)
+        end
+
+        @testset "Accuracy Test" begin
```
Comment (mofeing) [inline test/Numerics_test.jl:163]:
```suggestion
        @testset "[accuracy]" begin
```

**Review by mofeing — CHANGES_REQUESTED** [review]
Great! Just some aesthetic changes.

**Review by jofrevalles — COMMENTED** [review]
Thanks!

**Pull request merged**